### PR TITLE
Add principal converter

### PR DESCRIPTION
**Pull request by sophokles73**
The Principal instances created by Scandium when a client has been
authenticated contain only the authentication identifier used by the
client, e.g. the PSK identity or the client cert subject DN.

However, applications will often map the (technical) authentication
identifier used at the DTLS level to a logical application level
identifier which is used to refer to the client.

The PrincipalConverter can be implemented by applications to convert the
technical Principal created by Scandium to a Principal that contains
applications level information which can then be used when processing
CoAP requests.

### CODE REVIEW

**Comment by boaks** [comment]
My feeling is, that this would break the functionality of the `PrincipalSerializer`.
May be adding a `CustomPrincipalSerializer` to the config and  `PrincipalSerializer` fixes it.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java [hunk 1]
```diff
@@ -286,6 +287,8 @@
 	 */
 	private ConnectionIdGenerator connectionIdGenerator;
 
+	private PrincipalConverter principalConverter;
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java]: New fields must be included in the `protected Object clone()`.

**Comment by boaks** [comment]
The test covers, that the `convert`is called?
A test, which verifies, that the returned Principal is used in the `EndpointContext` would also be interested.

**Comment by sophokles73** [comment]
> A test, which verifies, that the returned Principal is used in the EndpointContext would also be interested.

will do

**Comment by sophokles73** [comment]
> My feeling is, that this would break the functionality of the PrincipalSerializer.

Good point. I'll take a look at the existing serializer and try to come up with some options ...

**Comment by boaks** [comment]
If the `PrincipalConverter` is moved into the "element-connector" module in the package `org.eclipse.californium.elements.auth`, it may be also used with TLS.

**Comment by boaks** [comment]
> My feeling is, that this would break the functionality of the PrincipalSerializer.

> Good point. I'll take a look at the existing serializer and try to come up with some options ...

I think, the `PrincipalSerializer` must be redesigned to remove the `static`s. 
If a custom Prinicpal is used, the serialization may use a ClientAuthenticationType.CUSTOM to differentiate between the standard and custom principals.

**Comment by boaks** [comment]
The new commit seems not to address the serialization issue, or?

File: element-connector/src/main/java/org/eclipse/californium/elements/auth/AbstractExtensiblePrincipal.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
+/**
```
Comment (boaks) [inline element-connector/src/main/java/org/eclipse/californium/elements/auth/AbstractExtensiblePrincipal.java]: Please replace that with the californium headers.
We may change that according EPL-2.0 after the 2.0 release.

Comment (sophokles73) [reply element-connector/src/main/java/org/eclipse/californium/elements/auth/AbstractExtensiblePrincipal.java]: sorry, settings from my Hono workspace keep getting in the way ...

**Comment by sophokles73** [comment]
> The new commit seems not to address the serialization issue, or?

Well, my hope is that we do not need to include the application specific info in serialization because the information is restored (using the `ApplicationLevelInfoSupplier` every time a session is established. I think I have managed to do it for the standard full and abbreviated handshake. However, I am not sure if it already is done in case of a fail-over between nodes ...

File: scandium-core/src/main/java/org/eclipse/californium/scandium/auth/ApplicationLevelInfoSupplier.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ */
+public interface ApplicationLevelInfoSupplier {
+
+    /**
+     * Gets additional information about an authenticated client.
+     * 
+     * @param clientIdentity The client identity.
+     * @return The additional information for the client.
+     */
+    Set<Principal> getInfo(Principal clientIdentity);
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/auth/ApplicationLevelInfoSupplier.java]: The `Set`-solution may also fit for issue #974 .

Is there any intention to access a specific `Principal` in `Set`?
Any thoughts about using a `Map<String, Principal>` ?

**Comment by boaks** [comment]
Except of the license header of the new files, LGTM.
If possible, please let us know, how you intend to operate on the Set or if a Map would do it better.

**Comment by sophokles73** [comment]
A Map would also work for me. We could probably also use a Map<String, Object> instead.

**Comment by boaks** [comment]
OK, so you didn't use the `Set` with a strong idea behind. I would prefer a `Map<String, Principal>`, as FMPOV this matches best for me.

**Comment by sophokles73** [comment]
ok, I'll make the corresponding changes...

**Comment by sophokles73** [comment]
@boaks I have pushed the changes ...

**Comment by boaks** [comment]
LGTM

**Comment by sbernard31** [comment]
The purpose of this PR is to avoid to "retrieve" each time application level identity from connection level identity, right ?

Why limited this feature to server side ?
Maybe the call of `amendClientPrincipal()` could be done on `sessionEstablished` ?

**Comment by sophokles73** [comment]
> The purpose of this PR is to avoid to "retrieve" each time application level identity from connection level identity, right ?

right

> Why limited this feature to server side ?

Just because we only need it on server side in Eclipse Hono. We only implement a CoAP server which never acts as a client initiating a connection.
However, this could be added in another PR later on (if somebody needs it) ...

**Comment by sbernard31** [comment]
To extends this features to both side server and client, I think it's mainly about : 
- naming : changing "client" by "foreign peer".
- moving amendPrincipal to `sessionEstablished()` in  handshaker ?

**Comment by sophokles73** [comment]
@sbernard31 I'll give it a try ...

**Comment by sophokles73** [comment]
@sbernard31 WDYT?

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java [hunk 4]
```diff
@@ -1118,4 +1125,28 @@ public void verifyCertificate(CertificateMessage message) throws HandshakeExcept
...
+	/**
+	 * Amends the peer principal with additional application level information.
+	 */
+	private void amendPeerPrincipal() {
+
+		Principal peerIdentity = session.getPeerIdentity();
+		if (peerIdentity instanceof ExtensiblePrincipal) {
+			// amend the client principal with additional application level information
+			@SuppressWarnings("unchecked")
+			ExtensiblePrincipal<? extends Principal> extensibleClientIdentity = (ExtensiblePrincipal<? extends Principal>) peerIdentity;
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java]: // amend the **peer** principal with additional application level information
@SuppressWarnings("unchecked")
ExtensiblePrincipal<? extends Principal> **extensibleIdentity** = (ExtensiblePrincipal<? extends Principal>) peerIdentity;

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ServerHandshaker.java [hunk 5]
```diff
@@ -56,16 +56,21 @@
...
 import java.util.List;
+import java.util.Map;
 
+import org.eclipse.californium.elements.auth.ExtensiblePrincipal;
 import org.eclipse.californium.elements.auth.PreSharedKeyIdentity;
 import org.eclipse.californium.elements.auth.RawPublicKeyIdentity;
 import org.eclipse.californium.elements.auth.X509CertPath;
 import org.eclipse.californium.elements.util.DatagramWriter;
 import org.eclipse.californium.elements.util.StringUtil;
+import org.eclipse.californium.scandium.auth.ApplicationLevelInfoSupplier;
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ServerHandshaker.java]: maybe some unused import in this class ?

File: scandium-core/src/test/java/org/eclipse/californium/scandium/ConnectorHelper.java [hunk 6]
```diff
@@ -278,6 +281,15 @@ LatchDecrementingRawDataChannel givenAnEstablishedSession(DTLSConnector client,
 		return clientChannel;
 	}
 
+	static void assertClientPrincipalHasAdditionalInfo(Principal clientIdentity, String expectedName) {
```
Comment (sbernard31) [inline scandium-core/src/test/java/org/eclipse/californium/scandium/ConnectorHelper.java]: assertPrincipalHasAdditionalInfo ?

**Comment by boaks** [comment]
@sophokles73 
Please rebase to "master" and create a new PR towards master.
The "conflict" is in the unit tests. Should be easy to fix. If not, remove my new tests to resolve these conflicts, I will then add them again after 2.0.0-M16

**Comment by sophokles73** [comment]
@boaks @sbernard31 I have created class `AdditionalInfo` which encapsuates a `Map<String, Object>` and provides a typed getter for the data. I have also rebased to master. Hope this works for everybody now :-)

**Comment by boaks** [comment]
The PR is towards 2.0.x, which is since last Friday, "history".
Please create it towards "master".

**Comment by boaks** [comment]
I checked the commits. they are already rebased on "master". Therefore the diff to 2.0.x get's large.
Please close this PR and open a new one towards master.

**Comment by sophokles73** [comment]
replaced by #1014

**Pull request closed without merging**